Move yeelight into component

zewelor opened the pull request:
## Breaking Changes

Move yeelight from light platform to component. As more yeelights are being released, with more features. It will make possible to support them. For example adding sensors for ceiling light, to get current power mode ( daylight / nightlight ). Or switches to support turning on off moonlight ( without having to use service call ). Also switch to turn off both main light and ambilight ( ceiling 650 ). Probably something more, as we will see what ideas we can make for better UX and support.

If you have following configuration:
```yaml
light:
  - platform: yeelight
    devices:
      192.168.1.13:
        name: Wall light
    custom_effects:
      - name: 'Fire Flicker'
        flow_params:
          count: 0
          transitions:
            - TemperatureTransition: [1900, 1000, 80]
            - TemperatureTransition: [1900, 2000, 60]
            - SleepTransition:       [1000]
```

It needs to be migrated to:
```yaml
yeelight:
  devices:
    192.168.1.13:
      name: Wall light
  custom_effects:
    - name: 'Fire Flicker'
      flow_params:
        count: 0
        transitions:
          - TemperatureTransition: [1900, 1000, 80]
          - TemperatureTransition: [1900, 2000, 60]
          - SleepTransition:       [1000]  
```

Just move everything to yeelight namespace, and remove 2 spaces vertical.

## Description:
Moved yeelight into component to allow adding switches / sensors etc. Now only option to change yeelight power mode ( daylight / nightlight in yeelight ceiling lamps ), is to call service. Then its not possible to see current power mode. Added nightlight switch to turn on / off nightlight mode and get feedback about currently enabled power mode. I'm not sure about many things, regarding component -> platform communication and setup. I've looked at other components and tried to come up with something works. Whole config is identical as it was for light platform. Looking to early feedback if its done right, then I will add docs update if its ok.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** https://github.com/home-assistant/home-assistant.io/pull/8845

## Example entry for `configuration.yaml` (if applicable):
```yaml
yeelight:
  devices:
    192.168.1.2:
      name: Yeelight
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

zewelor committed:
Move yeelight into component and add power mode switch

rytilahti commented:
I am going to try this out later on, but would it make sense to convert yeelight to use their custom protocol alongside with this conversion? This way we would gain the unique ids for the devices, allowing renames in the UI: https://yeelight.readthedocs.io/en/latest/#usage

zewelor commented:
Thanks for initial feedback. I will work on tests and fix lint issues. 

You mean id presented in discover_bulbs example ? Not sure where it should be used, I assume to somehow register devices with it ?

I was also wondering, how to approach light / switch synchonization of update. For example, when light is off, switching switch will turn it on, but then light waits up to 30s to update status. Is there any guideline how to synchonize between different platforms ? I was thinking light object could be passed to switch, but it looks like a bad coupling.

rytilahti commented:
Yeah, I'm talking about using that id as `unique_id` inside homeassistant's entity registry (https://developers.home-assistant.io/docs/en/entity_registry_index.html). This will allow homeassistant to identify the bulb uniquely even if its IP address changes and will enable the "rename device" option in the UI (so no need to manually use friendly_name mappings). As this information is not available with mDNS discovery (what is currently used), it has not been possible earlier.

Wrt syncing, I think you can use the `schedule_update` (https://developers.home-assistant.io/docs/en/entity_index.html#subscribing-to-updates) even when polling is used, but I hope someone will confirm that or propose another approache.

Wrt coupling, I think it's fine to pass a reference of the light to the switch as they are coupled anyway.

zewelor committed:
Synchronize state between entities, fixes for nightlight

zewelor commented:
I think I've figures out how to synchronize state between entities. Code taken from KNX integration.

Also I've put some fixes, regarding nightlight:
https://github.com/home-assistant/home-assistant/pull/21593/commits/d85bc75f19afd8d7a8c1e466647159849710954c#diff-6b4ac627753ee21cde162f000848ba3bR326

https://github.com/home-assistant/home-assistant/pull/21593/commits/d85bc75f19afd8d7a8c1e466647159849710954c#diff-6b4ac627753ee21cde162f000848ba3bR312

zewelor committed:
Lint fixes

zewelor commented:
Fixed lint errors.

I've also asked on discord about help with writing tests. Response I've got was that requirements for test is not written in checklist, so maybe they are not needed ( https://developers.home-assistant.io/docs/en/creating_component_code_review.html ). I've tried looking at some similar platform, to get idea how to approach / what to tests, but still not sure how to write them correctly.

As for unique id, not sure how to approach that ? Should discovery_bulb be done in component setup ? Or register some background job, and when response come, update all bulb devices to put received id ? Also is it needed to be done here. Maybe putting too large PR will make it harder to merge and somehow blocks other PR regarding yeelight. There is ongoing effort to add ambilight support for ceiling 650. Also I was thinking about adding some more config options, in separated PRs. Or its better to make one big ?

rytilahti commented:
I cannot say about the test requirements, but I had to create some when porting the tplink into a component.. I have not yet found time to test these changes and cannot currently promise when I'm able to do that, so I hope someone else chimes in. 

Anyway, on the `unique_id`s and other improvements, I think it's better to done it in phases to keep it simple so please ignore that for now. That conversion will be a bit more involved, basically requiring creating a config entry flow (https://developers.home-assistant.io/docs/en/config_entries_index.html) to do the discovery and initializing the platforms based on the discovery results.

zewelor committed:
Move yeelight services to component domain

zewelor commented:
Moved services into component domain. I think this is ready.

zewelor committed:
Remove switch from PR

zewelor commented:
@rytilahti Its ready, rebased on newest code. Can you take a look at it ? I've removed switch as I'm not sure about its UX ( What it should do on off etc). Will add sensor to get current working mode for daylight / nightlight for ceiling lights.

syssi reviewed (approved):
LGTM! Works fine with my Yeelight Color 1gen.

syssi commented:
Could you fix the conflict? Thanks!

zewelor commented:
@syssi Thanks, rebased on new code. Are you planning on adding ceiling4 ambilight support ? I saw some post about it on community forum. I was planning to work on it, don't want to duplicate work. After merging this I wanted to make PR with python-yeelight bump and some cleaning, that will be possible with new lib version. And then I can work on ambilight support. Also I want to add power mode sensor, for better UX with moonlight mode for ceilings.

syssi commented:
@zewelor I would be happy if you do most of the job. I don't own the device. I'd like to provide some support.

MartinHjelmare commented on homeassistant/components/yeelight/__init__.py:
```diff
@@ -1 +1,357 @@
+Support for Xiaomi Yeelight Wifi color bulb.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/yeelight/
```
We don't add the url in the docstring anymore.

MartinHjelmare commented on homeassistant/components/yeelight/__init__.py:
```diff
@@ -1 +1,357 @@
+        for device in yeelight_data[CONF_DEVICES].values():
+            device.update()
+
+    async_track_time_interval(
```
We can't  call an async function from sync context or vice versa. Use `track_time_interval` instead.

MartinHjelmare commented on homeassistant/components/yeelight/__init__.py:
```diff
@@ -1 +1,357 @@
+
+    discovery.listen(hass, SERVICE_YEELIGHT, device_discovered)
+
+    def async_update(event):
```
Why name this `async_update` when it's not a coroutine or callback decorated function?

MartinHjelmare commented on homeassistant/components/yeelight/light.py:
```diff
@@ -160,117 +84,39 @@ def _wrap(self, *args, **kwargs):
-    hass.services.register(
-        DOMAIN, SERVICE_START_FLOW, service_handler,
-        schema=service_schema_start_flow)
+    yeelight_data[CONF_LIGHTS][ipaddr] = light
```
We're not allowed to store the entity itself in a shared container. The entity is a platform concern. It's ok to store a reference to the entity, eg the entity_id.

MartinHjelmare commented on homeassistant/components/yeelight/__init__.py:
```diff
@@ -1 +1,357 @@
+        entity_ids = extract_entity_ids(hass, service)
+        target_devices = [dev.device for dev in
+                          yeelight_data[CONF_LIGHTS].values()
+                          if dev.entity_id in entity_ids]
```
We don't need to access the entity here. It's enough to check for the correct entity_ids. See comment below about storing entities vs entity_ids.

MartinHjelmare reviewed (commented):
Please open a new PR where we can address the comments.

zewelor commented:
@MartinHjelmare PR with fixes here: https://github.com/home-assistant/home-assistant/pull/22347

sergeymaysak commented:
@zewelor First, I really appreciate this effort which significantly improves usage of my yeelight 650. I've tested today in merged to dev branch. Yes - since now it allows me to control ambient lights as well. Great job, really! However there are several issues which I believe important to depict and possibly to fix maybe i next releases of this great contribution:
1. the most significant issue imo - if my configuration.yaml does not have any notions of yeelight platform it still discovered by hass (Bonjour/mDNS is here ;) for good) and before this contribution I had just auto-discovered yeelight light representing main bulb (yes, ambient lighs missed). With this contribution I have exception while hass load:
```
2019-03-27 20:12:15 ERROR (MainThread) [homeassistant.setup] Error during setup of component yeelight
Traceback (most recent call last):
  File "/Users/sam/0.homeassistant/home-assistant/homeassistant/setup.py", line 154, in _async_setup_component
    component.setup, hass, processed_config)  # type: ignore
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/sam/0.homeassistant/home-assistant/homeassistant/components/yeelight/__init__.py", line 147, in setup
    conf = config[DOMAIN]
KeyError: 'yeelight'
```
and no lights at all. 
So yes, after addition of `yeelight:` into configuration.yaml its autodiscovered my lights and now its shows two lights for device - main and ambient! great! as expected.
But the point is that overall user experience is completely suffered - instead of `magical` auto-discovered device which is really shows the power and elegance and magic of HA to end users - users would have to specify platform explicitly. Pitty. If hovewer `KeyError` whould be just suppressed in code it would provide significant improvement for user experience. Please consider. And yes - I do not specify my yeelight ip addresses - yes I want Bonjour and auto-discovery do it for me. Yes I do really dont want to modify my configuration.yaml for cases that do not introduce anything specific - I love when my new devices are auto-discovered ...
2. issue number two but less significant ):
which this contribution main light lost ability to change temperature and specify effect from UI:
<img width="366" alt="Screen Shot 2019-03-27 at 11 18 47 PM" src="https://user-images.githubusercontent.com/11465932/55113069-d5223600-50e6-11e9-8497-9d18586cf17d.png">

So only brightness can be changed. Well, hmmm - presence of ability to control ambient light is really more important for me, but sounds as defect imo...

Thanks again for overall great contribution - hope you would share my vision on UX here ) Cheers!

MartinHjelmare commented:
Please open an issue if you suspect a bug.

If you want to suggest an enhancement please open a feature request in [the Feature Requests section](https://community.home-assistant.io/c/feature-requests) of our community forum.

Merged PRs should not be used for enhancement discussion or bug reports. If you've found a bug it's ok to make a review with inline comments and link to an issue that reports the bug.

Thanks!
